File: src/p2xxx/p2054.py
from leetcode_prelude import *


# Problem 2054
class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        timeline = []

        for start_time, end_time, score in events:
            timeline.append((start_time, 1, score))
            timeline.append((end_time + 1, 0, score))

        timeline.sort()

        res = 0
        best_finished = 0

        for _, is_start, score in timeline:
            if is_start:
                res = max(res, score + best_finished)
            else:
                best_finished = max(best_finished, score)

        return res

File: src/p2xxx/p2054.rs
// Problem 2054
impl Solution {
    pub fn max_two_events(mut events: Vec<Vec<i32>>) -> i32 {
        use std::cmp::Reverse;
        use std::collections::BinaryHeap;

        events.sort_by_key(|s| s[0]);

        let mut heap = BinaryHeap::new();

        let mut best = 0;
        let mut res = 0;

        for row in events.iter() {
            let &[start_time, end_time, value] = row.as_slice() else {
                panic!("bad format");
            };

            while let Some((Reverse(time), value)) = heap.peek().copied() {
                if time >= start_time {
                    break;
                }

                best = best.max(value);
                heap.pop();
            }

            res = res.max(value + best);

            heap.push((Reverse(end_time), value));
        }

        res
    }
}
